Update v4 migration documentation (#2236)
graphql-dotnet/graphql-dotnet: docs2/site/docs/migrations/migration4.md

    {
        Name = "HumanInput";
        Field<NonNullGraphType<StringGraphType>>("name");
        Field<StringGraphType>("homePlanet");
    }

    public override object ParseDictionary(IDictionary<string, object> value)
    {
        return new Human
        {
            Name = ((string)value["name"]).ToUpper(),
            HomePlanet = value.TryGetValue("homePlanet", out var homePlanet) ? (string)homePlanet : "Unknown",
            Id = null,
        };
    }
}
```

Note that pursuant to GraphQL specifications, if a field is optional, not supplied, and has no default,
it will not be in the dictionary.

For untyped `InputObjectGraphType` classes, like shown above, the default behavior of `ParseDictionary`
will be to return the dictionary. `GetArgument<T>` will still attempt to convert a dictionary to the
requested type via `ObjectExtensions.ToObject` as it did before.

### Experimental Features / Applied Directives

> Ability to apply directives to the schema elements and expose user-defined meta-information
  via introspection - `schema.EnableExperimentalIntrospectionFeatures()`.
> See https://github.com/graphql/graphql-spec/issues/300 for more information.

> (sungram3r todo: write separate page on these features, write a simple summary and reasoning here, and add link to new page)

Reviewer: ```suggestion
(sungam3r todo: write separate page on these features, write a simple summary and reasoning here, and add link to new page)
```

### Ability to Sort Introspection Results

Introspection results are now sorted based on a configured 'comparer' for a schema. You can configure the comparer by setting
`ISchema.Comparer` to an implementation of `ISchemaComparer`. By default, introspection results are returned in the order they
were defined.

See [Default Sort Order of Introspection Query Results](#Default-Sort-Order-of-Introspection-Query-Results) below for a sample
of how this can be used to return introspection results that are sorted alphabetically.

### Array Pooling

When returning lists of information from field resolvers, you can choose to rent an array from `IResolveFieldContext.ArrayPool`,
populating it with your results and returning the array. The array will be released after the execution completes. This has
limited uses, since the rented array is not guaranteed to be exactly the requested length, so the array would need to be

> wrapped in order to only return the correct number of entries, triggering a memory allocation (albeit a smaller one).

Reviewer: ```suggestion
wrapped in order to only return the correct number of entries, triggering a memory allocation (albeit a smaller one):
```c#
 resolve: context =>
{
    var ints = context.ArrayPool.Rent<int>(1000); // ints.Length >= 1000
    for (int i=0; i<1000; ++i)
        ints[i] = i;
    return ints.Constrained(1000); // extension method to return an array or array-like object of a given length
});
```
```

Reviewer: > array would need to be wrapped 

I have revealed the meaning.

- `EnableReadDescriptionFromAttributes` enables or disables setting default values for 'description' from `DescriptionAttribute`. Enabled by default.
- `EnableReadDescriptionFromXmlDocumentation` enables or disables setting default values for 'description' from XML documentation. Disabled by default.
- `Validation` configures the validator used when setting the `Name` property on types, arguments, etc. Can be used to disable validation
  when the configured `INameConverter` fixes up invalid names. See `ISchema.NameConverter`.
- `ValidationOnSchemaInitialize` configures the validator used to verify the schema after the `INameConverter` has processed all the names.
  Disabling this validator is unlikley to be of any use, since the parser will not be able to parse a document that contains invalid characters in a name.

It is recommended to configure these options once when your application starts, such as within your `void Main()` method, a static
constructor of your schema, or a similar location.

### Authorization Extension Methods

> Extension methods to configure authorization requirements for GraphQL elements: types, fields, schema.

> (sungram3r todo: write more description about how it interacts with the other libraries, add simple sample)

Reviewer: ```suggestion
(sungam3r todo: write more description about how it interacts with the other libraries, add simple sample)
```

document caching.

To facilitate the performance changes, many changes were made to the API that may affect you if you have
built custom execution strategies, scalars, parser, or similar core components. Please see the complete list
of breaking changes below.

### Input Object Custom Deserializers (aka resolver)

You can now add code to `InputObjectGraphType` descendants to build an object from the collected argument
fields. The new `ParseDictionary` method is called when variables are being parsed or `GetArgument` is called,
depending on if the argument is stored within variables or as a literal. The method is passed a dictionary
containing the input object's fields and deserialized values.

By default, for `InputObjectGraphType<TSourceType>` implementations, the dictionary is passed to
`ObjectExtensions.ToObject` in order to convert the dictionary to an object of `TSourceType`.
You can override the method to have it return an instance of any appropriate type.

Below is a sample which sets a default value for an unsupplied field (this could be done with a default
value set on the field, of course) and converts the name to uppercase:

> ```cs

Reviewer: c#

Author: ```suggestion
```csharp
```

`ISchema.Comparer` to an implementation of `ISchemaComparer`. By default, introspection results are returned in the order they
were defined.

See [Default Sort Order of Introspection Query Results](#Default-Sort-Order-of-Introspection-Query-Results) below for a sample
of how this can be used to return introspection results that are sorted alphabetically.

### Array Pooling

When returning lists of information from field resolvers, you can choose to rent an array from `IResolveFieldContext.ArrayPool`,
populating it with your results and returning the array. The array will be released after the execution completes. This has
limited uses, since the rented array is not guaranteed to be exactly the requested length, so the array would need to be
wrapped in order to only return the correct number of entries, triggering a memory allocation (albeit a smaller one):
```c#
 resolve: context =>
{
    var ints = context.ArrayPool.Rent<int>(1000); // ints.Length >= 1000
    for (int i=0; i<1000; ++i)
        ints[i] = i;
    return ints.Constrained(1000); // extension method to return an array or array-like object of a given length

> });

Reviewer: ```suggestion
});
```
```

> });

Reviewer: "```" lost

Author: ```suggestion
});
```
```

> It is not recommended to use this feature for interim calculations, as it is better to work directly with `System.Buffers.ArrayPool<T>`.

Reviewer: ```suggestion
```
It is not recommended to use this feature for interim calculations, as it is better to work directly with `System.Buffers.ArrayPool<T>`.
```

- `EnableReadDescriptionFromXmlDocumentation` enables or disables setting default values for 'description' from XML documentation. Disabled by default.
- `Validation` configures the validator used when setting the `Name` property on types, arguments, etc. Can be used to disable validation
  when the configured `INameConverter` fixes up invalid names. See `ISchema.NameConverter`.
- `ValidationOnSchemaInitialize` configures the validator used to verify the schema after the `INameConverter` has processed all the names.
  Disabling this validator is unlikley to be of any use, since the parser will not be able to parse a document that contains invalid characters in a name.

It is recommended to configure these options once when your application starts, such as within your `void Main()` method, a static
constructor of your schema, or a similar location.

### Authorization Extension Methods

> Extension methods to configure authorization requirements for GraphQL elements: types, fields, schema.

(sungam3r todo: write more description about how it interacts with the other libraries, add simple sample)

### Other Features

* New method `IParentExecutionNode.ApplyToChildren`

> * Support for repeatable directives and ability to expose `isRepeatable` field via introspection - `schema.ExperimentalFeatures.RepeatableDirectives`.

Reviewer: Not full support yet :(

* Schema validation upon initialization and better support for schema traversal via `ISchemaNodeVisitor`

## Breaking Changes

### Schema Configuration Options Moved

`NameConverter`, `SchemaFilter` and `FieldMiddleware` have been removed from `ExecutionOptions` and are now properties on `Schema`.
These properties can be set in the constructor of the `Schema` instance, or within your DI composition root, or at any time before
any query is executed. Once a query has been executed, changes to these fields is not allowed, and adding middleware via the field middleware
builder has no effect.

### Middleware Builders

- The signature of `IFieldMiddlewareBuilder.Use` has been changed to remove the schema from delegate. Since the schema is now known, there is no
  need for it to be passed to the middleware builder.
- The middleware `Use<T>` extension method has been removed. Please use the `Use` method with a middleware instance instead.

> See [Field Middleware](https://graphql-dotnet.github.io/docs/getting-started/field-middleware) for more information.

Reviewer: This page needs fixing.

Author: What's wrong with it?  It was updated inside #2210 along with the relevant changes.

Reviewer: I forgot to switch to `develop`. Everything is good but link to directives in https://github.com/graphql-dotnet/graphql-dotnet/blob/develop/docs2/site/docs/getting-started/field-middleware.md#field-middleware-vs-directive section is broken.

Author: Good catch.  Should probably update that with the rest of your changes to directive documentation.

use this instance once the resolver finishes executing. To preserve a copy of the context, call `.Copy()` on the context
to create a copy that is not re-used. Note that it is safe to use the field context within asynchronous field resolvers and
data loaders. Once the asynchronous field resolver or data loader returns its final result, the context will be cleared and may be re-used.
Also, any calls to the configured `UnhandledExceptionDelegate` will receive a field context copy that will not be re-used,
so it is safe to preserve these instances without calling `.Copy()`.

### Subscriptions Moved to Separate Project

The implementation for subscriptions, contained within `SubscriptionExecutionStrategy`, has been moved into the
[`GraphQL.SystemReactive`](https://www.nuget.org/packages/GraphQL.SystemReactive/) NuGet package. The default document executer
will now throw a `NotSupportedException` when attempting to execute a subscription. Please import the NuGet package and use the
`SubscriptionDocumentExecuter` instead. If you have a custom document executer, you can override `SelectExecutionStrategy` in
order to select the `SubscriptionExecutionStrategy` instance for subscriptions.

> ```cs

Reviewer: c# or csharp

Author: ```suggestion
```csharp
```

* `TypeInfo.GetAncestors()` has been changed to `TypeInfo.GetAncestor(int index)`
* Various methods within `StringUtils` have been removed; please use extension methods within `StringExtensions` instead.
* `ExecutionHelper.GetVariableValue` has been removed, and the signature for `ExecutionHelper.CoerceValue` has changed.
* Removed `TypeExtensions.As`
* `ExecutionHelper.CollectFields` method was moved into `Fields` class and renamed to `CollectFrom`
* `ISchema.FindDirective`, `ISchema.RegisterDirective`, `ISchema.RegisterDirectives` methods were moved into `SchemaDirectives` class
* `ISchema.FindType` method was moved into `SchemaTypes[string typeName]` indexer
* Some of the `ISchemaNodeVisitor` methods have been changes to better support schema traversal
* `SourceLocation`, `NameNode` and `BasicVisitor` have been changed to a `readonly struct`.
* `ObjectExtensions.GetInterface` has been removed along with two overloads of `GetPropertyValue`.
* `void INode.Visit<TState>(System.Action<INode, TState> action, TState state)` method has been added.
* Various `IEnumerable<T>` properties on schema and graph types have been changed to custom collections:

>   `SchemaDirectives`, `SchemaTypes`, `TypeFields`, `PossibleTypes`, and `ResolvedInterfaces`

Reviewer: ```suggestion
  `SchemaDirectives`, `SchemaTypes`, `TypeFields`, `PossibleTypes`, `Interfaces` and `ResolvedInterfaces`
```

did in 3.x. To re-enable that functionality, see [Global Switches](#Global-Switches) above.

### Changes to `IResolveFieldContext.Arguments`

`IResolveFieldContext.Arguments` now returns an `IDictionary<string, ArgumentValue>` instead of `IDictionary<string, object>` so that it
can be determined if the value returned is a default value or if it is a specified literal or variable.

`IResolveFieldContext.HasArgument` now returns `false` when `GetArgument` returns a field default value. Note that if a variable is specified,
and the variable resolves to its default value, then `HasArgument` returns `true` (since the field argument has successfully resolved to a variable
specified by the query).

### Metadata is Not Thread Safe

`IProvideMetadata.Metadata` is now a `Dictionary<string, object>` instead of `ConcurrentDictionary<string, object>`, and is not thread safe anymore.
If you need to write metadata during execution of field resolvers, lock on the graph type before accessing the dictionary. Do not lock on the
`Metadata` property because there can be concurrency issues accessing the field.

> ```cs

Reviewer: c#

Author: ```suggestion
```csharp
```

{
    int value;
    if (field.Metadata.TryGetValue("counter", out var valueObject)) value = (int)valueObject;
    field.Metadata["counter"] = value + 1;
}
```

### Other Breaking Changes

* GraphQL.NET now uses GraphQL-Parser v7 with new memory model taking advantage of `System.Memory` APIs.
* When used, Apollo tracing will now convert the starting timestamp to UTC so that `StartTime` and `EndTime` are properly serialized as UTC values.
* `Connection<TNode, TEdge>.TotalCount` has been changed from an `int` to an `int?`. This allows for returning `null` indicating that the total count is unknown.
* `InputObjectGraphType.ParseDictionary` has been added so that customized deserialization behavior can be specified for input objects (aka input resolvers).
  If `InputObjectGraphType<T>` is used, and `GetArgument<T>` is called with the same type, no behavior changes will occur by default.
  If `InputObjectGraphType<T>` is used, but `GetArgument<T>` is called with a different type, coercion may fail. Override `ParseDictionary`
  to force resolving the input object to the correct type. See [Input Object Custom Deserializers](#Input-Object-Custom-Deserializers-aka-resolver) above.

> * `ExecutionResult.Data` format breaking changes. Now objects are serialized as `ObjectProperty` arrays, not object dictionaries.

Reviewer: ```suggestion
* `ExecutionResult.Data` format breaking changes.
```

## New Features

### Improved Performance

> GraphQL.NET 4.0 has been highly optimized, typically executing queries at least 50% faster while also providing a 65% memory reduction. Small queries have been measured to run as much as twice as fast as previously. A cached query executor is also provided, which can reduce execution time another 20% once the query has been parsed (disabled by default). Variable parsing is also improved to run about 50% faster, and schema build time is now about 20x faster than previously and requires 1/25th the amount of memory.

Reviewer: > 1/25th the amount of memory

Where does this data come from?

Author: I ran tests just now.

Reviewer: 😮 😮 😮

Author: 100,000 bytes vs 2.6mb

Author: Disabling xml comments was huge

Author: I’ll figure out the merge tool and post the results when I have time

> GraphQL.NET 4.0 has been highly optimized, typically executing queries at least 50% faster while also providing a 65% memory reduction. Small queries have been measured to run as much as twice as fast as previously. A cached query executor is also provided, which can reduce execution time another 20% once the query has been parsed (disabled by default). Variable parsing is also improved to run about 50% faster, and schema build time is now about 20x faster than previously and requires 1/25th the amount of memory.

Reviewer: ```suggestion
GraphQL.NET 4.0 has been highly optimized, typically executing queries at least 50% faster while also providing a 65% memory reduction. Small queries have been measured to run as much as twice fast than previously. A cached query executor is also provided, which can reduce execution time another 20% once the query has been parsed (disabled by default). Variable parsing is also improved to run about 50% faster, and schema build time is now about 20x faster than previously and requires 1/25th the amount of memory.
```
?

Also please break long lines.

Author: "twice as fast as" is correct... but we could either make it shorter

"measured to run ~as much as~ twice as fast as previously"

and at the end, "previously" could instead be written as "the previous version" or "they previously ran".

"measured to run as much as twice as fast as ~previously~ the previous version."

"measured to run as much as twice as fast as ~previously~ they previously ran."

```suggestion
GraphQL.NET 4.0 has been highly optimized, typically executing queries at least 50% faster while also providing
a 65% memory reduction. Small queries have been measured to run twice as fast as they previously ran. A cached
query executor is also provided, which can reduce execution time another 20% once the query has been parsed
(disabled by default). Variable parsing is also improved to run about 50% faster, and schema build time is now
about 20x faster than previously and requires 1/25th the amount of memory.
```

Don't worry about the line breaks.  It's still a WIP and I'll get it all cleaned up.  I only just started on this.

> > Document caching supported via `IDocumentCache` and a default implementation within `DefaultDocumentCache`.

Reviewer: Are all quotes a placeholder?

Author: yes; they are just copies of what was there before.  I may need need to rewrite them, or just delete them once I'm happy with whatever I replaced it with -- or possibly just leave them.

> Within the `GraphQL.Caching` nuget package, a memory-backed implementation is available which is backed by `Microsoft.Extensions.Caching.Memory.IMemoryCache`.

To enable cached queries...(todo: add sample)

To facilitate the performance changes, many changes were made to the API that may affect you if you have built custom execution strategies, scalars, parser, or similar core components. Please see the complete list of breaking changes below.

### Input Object Custom Deserializers (aka resolver)

> You can now add code to `InputObjectGraphType` derivatives to build an object from the collected 

Reviewer: ```suggestion
You can now add code to `InputObjectGraphType` descendants to build an object from the collected arguments. 
```

    }
```

(todo: add sample line of code of how to set the comparer in the schema)

### `IResolveFieldContext` Re-use

The `IResolveFieldContext` instance passed to field resolvers is re-used at the completion of the resolver. Be sure not to
use this instance once the resolver finishes executing. To preserve a copy of the context, call `.Copy()` on the context
to create a copy that is not re-used. Note that it is safe to use the field context within asynchronous field resolvers and

> data loaders. Once the asynchronous field resolver or data loader returns its final result, the context may be re-used.

Reviewer: ```suggestion
data loaders. Once the asynchronous field resolver or data loader returns its final result, the context will be cleared and may be re-used.
```

Also, any calls to the configured `UnhandledExceptionDelegate` will receive a field context copy that will not be re-used,
so it is safe to preserve these instances without calling `.Copy()`.

### Subscriptions Moved to Separate Project

> Subscriptions implementation (`SubscriptionExecutionStrategy`) has been moved into `GraphQL.SystemReactive` project and default document executer throws `NotSupportedException`.

(todo: revise description with reference to proper nuget package, describe changes required to project, etc)

### API Cleanup

* `GraphQL.Instrumentation.StatsReport` and its associated classes have been removed. Please copy the source code into your project if you require these classes.
* `LightweightCache.First` method has been removed.
* `IGraphType.CollectTypes` method has been removed.
* `ExecutionHelper.SubFieldsFor` method has been removed.
* `NodeExtensions`, `AstNodeExtensions` classes have been removed.
* `CoreToVanillaConverter` class became `static` and most of its members have been removed.
* `GraphQL.Language.AST.Field.MergeSelectionSet` method has been removed.
* `CoreToVanillaConverter.Convert` method now requires only one `GraphQLDocument` argument.

> * `TypeCollectionContext` class is now internal, also all methods with this parameter in `GraphTypesLookup` are private.

Reviewer: no more `GraphTypesLookup`

* `TypeInfo.GetAncestors()` has been changed to `TypeInfo.GetAncestor(int index)`
* Various methods within `StringUtils` have been removed; please use extension methods within `StringExtensions` instead.
* `GraphTypesLookup` has been renamed to `SchemaTypes` with a significant decrease in public APIs 
* `ExecutionHelper.GetVariableValue` has been removed, and the signature for `ExecutionHelper.CoerceValue` has changed.
* Removed `TypeExtensions.As`
* `ExecutionHelper.CollectFields` method was moved into `Fields` class and renamed to `CollectFrom`
* `ISchema.FindDirective`, `ISchema.RegisterDirective`, `ISchema.RegisterDirectives` methods were moved into `SchemaDirectives` class
* `ISchema.FindType` method was moved into `SchemaTypes[string typeName]` indexer
* Some of the `ISchemaNodeVisitor` methods have been changes to better support schema traversal
* `SourceLocation`, `NameNode` and `BasicVisitor` have been changed to a `readonly struct`.
* `ObjectExtensions.GetInterface` has been removed along with two overloads of `GetPropertyValue`.

> * `void Visit<TState>(System.Action<INode, TState> action, TState state)` method has been added.

Reviewer: ```suggestion
* `void INode.Visit<TState>(System.Action<INode, TState> action, TState state)` method has been added.
```

* Various `IEnumerable<T>` properties on schema and graph types have been changed to custom collections: `SchemaDirectives`, `SchemaTypes`, `TypeFields`, `PossibleTypes`, `ResolvedInterfaces`
* `INode.IsEqualTo` and related methods have been removed.
* `ApolloTracing.ConvertTime` is now private and `ResolverTrace.Path` does not initialize an empty list when created.

### `ExecutionOptions.EnableMetrics` is disabled by default

> To enable metrics, please set the option to `true` before executing the query.

Reviewer: Also see this https://github.com/graphql-dotnet/server/issues/498

(todo: add sample)

### GraphQL Member Descriptions

> By default, descriptions for fields, types, enums, and so on are not pulled from xml comments unless the corresponding global flag is enabled.

> Note that to improve performance, by default GraphQL.NET 4.0 does not pull descriptions from fields from xml comments as it did in 3.x. To re-enable that functionality, see Global Switches above.

Reviewer: ```suggestion
Note that to improve performance, by default GraphQL.NET 4.0 does not pull descriptions for fields from xml comments as it did in 3.x. To re-enable that functionality, see Global Switches above.
```

(todo: update/consolidate description)

### Changes to `IResolveFieldContext.Arguments`

`IResolveFieldContext.Arguments` now returns an `IDictionary<string, ArgumentValue>` instead of `IDictionary<string, object>` so that it
can be determined if the value returned is a default value or if it is a specified literal or variable.

`IResolveFieldContext.HasArgument` now returns `false` when `GetArgument` returns a field default value. Note that if a variable is specified,

> and the variable resolves to its default value, then `HasArgument` returns `true` (since the field argument is successfully resolving to a variable

Reviewer: ```suggestion
and the variable resolves to its default value, then `HasArgument` returns `true` (since the field argument is successfully resolved to a variable
```

?

specified by the query).

### Metadata is Not Thread Safe

> `IProvideMetadata.Metadata` is now `Dictionary` instead of `ConcurrentDictionary`, and is not thread safe anymore

(todo: update description, add description of how to lock on type, not dictionary)

### Other Breaking Changes

> * GraphQL.NET now uses GraphQL-Parser v7 with new memory model

Reviewer: ```suggestion
* GraphQL.NET now uses GraphQL-Parser v7 with new memory model taking advantage of `System.Memory` APIs
```

* When used, Apollo tracing will now convert the starting timestamp to UTC so that `StartTime` and `EndTime` are properly serialized as UTC values.
* `Connection<TNode, TEdge>.TotalCount` has been changed from an `int` to an `int?`. This allows for returning `null` indicating that the total count is unknown.

> * `InputObjectGraphType.ParseDictionary` has been added so that customized deserialization behavior can be specified for input objects.

Reviewer: ```suggestion
* `InputObjectGraphType.ParseDictionary` has been added so that customized deserialization behavior can be specified for input objects (aka input resolvers).
```
